### PR TITLE
Feat: HTML spatial mapping

### LINKED ISSUE #18: [Feat]: HTML spatial mapping
### Feature Type

Adding new functionality

### Problem Description

Producing maps which are interactive and can have multiple layers is complex and not standardised. Implement the HTML mapping from bronze-offer-automation-tools would allow for a simple consistent approach.

### Impact

Use of static maps which aren't as easy for others to use.

### Acceptance Criteria

- New module which uses folium to output standalone HTML maps with multiple layers.
- Easy to use API with minimal inputs to produce a map e.g. GeoDataFrame and column for colouring
- Provide multiple GeoDataFrames to have separate (toggleable) layers on a map
- Options to provide more specific colouring, styling and legends

Some nice to haves:
- Detailed documentation on the functions / classes with examples
- More detailed standalone code examples
- Ability to produce a HTML map from CLI which takes a config to define the layers and parameters

### Proposed Approach

Copy [mapping module from bronze-offer-automation-tools](https://github.com/Transport-for-the-North/bronze-offer-automation-tools/blob/main/src/bronze_offer_automation_tools/mapping.py) and tweak it where necessary to make it more general and user friendly.

### Alternative Solutions

_No response_

### Additional Context

Speak to @amber-mertens-tfn for more details on the BOATs implementation.

### PR DESCRIPTION
**Pull request by amber-mertens-tfn**
# Changes

Adding new html mapping functionality:
Need to update the example to also have split maps. Also for some reason it doesn't center on the data anymore, need to fix that.

- Closes #XXXX *(replace XXXX with GitHub issue number)*
- ...

# Tasks

- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Has documentation (including user guide) been updated
- [x] Have new dependencies been added (or changed) in relevant `requirements.txt`
- [x] Code linting, tests and other workflows pass

### CODE REVIEW

File: examples/html_mapping_example.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+"""
+Example
+===========
+
+This is a code example which is displayed but **not** run.
```
Comment (MattBuckley-TfN) [inline examples/html_mapping_example.py]: Please could you update the title and description of the example.

File: examples/html_mapping_example.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+from caf.viz.web import mapping
+
+# %%
+# CONSTANTS #
+COUNTRY_CODE = 0
+REGION_CODE = 3
+
+# PATHS
+PATH_TO_SAVE_MAP = pathlib.Path(r"path/to/save/map.html")
+PATH_TO_SAVE_SPLIT_MAP = pathlib.Path(r"path/to/save/split_map.html")
```
Comment (MattBuckley-TfN) [inline examples/html_mapping_example.py]: Update these so that the script can be ran without making any changes, it will fail to save files because those folders don't exist but will be fine to save them to the working directory.
```suggestion
PATH_TO_SAVE_MAP = pathlib.Path(r"map.html")
PATH_TO_SAVE_SPLIT_MAP = pathlib.Path(r"split_map.html")
```

File: examples/html_mapping_example.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
+"""
+Example
+===========
+
+This is a code example which is displayed but **not** run.
+
+"""
+
+# %%
+# IMPORTS #
```
Comment (MattBuckley-TfN) [inline examples/html_mapping_example.py]: In these example files any Python comments at the top of the cell block, immediately after `# %%`, are treated as markdown (or restructured text) and displayed fully in the example. This means you can use headings (or other text styling) e.g.
```suggestion
# %%
# Imports
# -------
```

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
New module looks good, just a couple of comments to add more detail to the example.

Also, the example isn't being ran when generating the docs, if you wanted to have it ran so the outputs were displayed in the docs the filename can be changed to start with "run_" e.g "run_html_mapping_example.py". I'm not sure how well the outputs will embed within the docs for folium but it does work really well for static plots from matplotlib and I think should work for folium but might involve displaying the map directly instead of saving it.

If you want to see the generated documentation you can call `make html` from within the docs folder or look at the [PR deployed on read the docs](https://cafviz--32.org.readthedocs.build/en/32/_generated/examples/html_mapping_example.html)

File: examples/run_html_mapping_example.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+===========
+
+This is a code example showing how to create an interactive map using the web/mapping module.
+This example shows two ways to create an interactive html map:
+A single map with the desired datasets,and a split map consisting of an overview map with the
+ split geometries which link to individual maps showing the datasets for each split geometry.
+
+The example uses the NUTS dataset from Eurostat and the cities dataset from Natural Earth,
+ both of which are available in geodatasets.
+
```
Comment (MattBuckley-TfN) [inline examples/run_html_mapping_example.py:12]: Remove the indentation at the start of the lines because it causes formatting issues in the HTML.
```suggestion
A single map with the desired datasets,and a split map consisting of an overview map with the
split geometries which link to individual maps showing the datasets for each split geometry.

The example uses the NUTS dataset from Eurostat and the cities dataset from Natural Earth,
both of which are available in geodatasets.

```

File: requirements.txt [hunk 5]
```diff
@@ -4,3 +4,8 @@ pandas>=1.4.0
 matplotlib>=3.8.2
 scipy>=1.11.4
 pydantic>=2.0.0
+folium>=0.20.0
+geopandas>=1.1.3
+tqdm>=4.67.3
+branca>=0.8.2
+shapely>=2.1.2
```
Comment (MattBuckley-TfN) [inline requirements.txt:11]: mapclassify is required for producing the heatmaps with explore
```suggestion
shapely>=2.1.2
mapclassify>=2.10
```

File: examples/run_html_mapping_example.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+# ------------------------
+m = mapping.map_datasets(datasets=mapping_datasets, mask=filter_zones, mask_name="Europe")
+
+# %%
+m  # noqa: B018
+
+# %%
+# SAVE MAP (OPTIONAL)
+# ------------------------
+m.save(PATH_TO_SAVE_MAP)
```
Comment (MattBuckley-TfN) [inline examples/run_html_mapping_example.py]: To simplify this example I wouldn't bother actually storing the map object as a variable or saving instead I would just explain that.
```suggestion
# %%
# Create a Single Map
# -------------------
# :func:`~caf.viz.mapping.map_datasets` will create a :class:`folium.Map` object from the datasets with
# OpenStreetMap background, it can be saved to a standalone HTML file with :meth:`folium.Map.save`.
mapping.map_datasets(datasets=mapping_datasets, mask=filter_zones, mask_name="Europe")
```

File: examples/run_html_mapping_example.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+# %%
+# CREATE MAP (split map)
+# ------------------------
```
Comment (MattBuckley-TfN) [inline examples/run_html_mapping_example.py]: Update the output path to it's saved to the generated HTML docs and add a link in the description
```suggestion
# %%
# Create a Split Map
# ------------------
# Create a map which is split across multiple HTML files to allow for more detailed
# data on each subset map, see `Split Map Overview <split_map.html>`_.
#
# .. note::
#     HTML file is written to generated documentation folder, so it's deployed with
#     documentation, normally this can be saved anywhere.
PATH_TO_SAVE_SPLIT_MAP = pathlib.Path(r"../docs/build/html/_generated/examples/split_map.html")
PATH_TO_SAVE_SPLIT_MAP.parent.mkdir(exist_ok=True, parents=True)
```

File: examples/run_html_mapping_example.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+# %%
+# CONSTANTS
+# ----------------
+COUNTRY_CODE = 0
+REGION_CODE = 3
+
+# PATHS
+PATH_TO_SAVE_MAP = pathlib.Path(r"map.html")
+PATH_TO_SAVE_SPLIT_MAP = pathlib.Path(r"split_map.html")
```
Comment (MattBuckley-TfN) [inline examples/run_html_mapping_example.py]: Not needed if not saving the map, from [this comment](https://github.com/Transport-for-the-North/caf.viz/pull/32#discussion_r3387106879). Also, I think the split map path should be moved to nearer the place it's used see [this other comment](https://github.com/Transport-for-the-North/caf.viz/pull/32#discussion_r3387318495).
```suggestion
```

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
This example looks great, just a few last comments.

A few to tweak the wording of the example, also please could you change the section titles so they aren't all uppercase and add a sentence describing what the code block underneath is doing, similar to the changes I've suggested for create map sections.

Also, an `__init__.py` file should be added to `src\caf\viz\web` so that Python knows web is a sub-package not just a folder. The `__init__.py` can be empty but it would be good to add a docstring which explains the purpose of the web sub-package.

**Review by MattBuckley-TfN — APPROVED** [review]
Looks great thanks for that, there seems to be an issue with the link to the split maps but I'll have a look into than in another branch because I think it's something to do with how read the docs produces the outputs.
